chi2/chinotype.py

```diff
@@ -380,23 +413,67 @@ def prepChi(self):
                     group by ccd
                 ) chicon
                 left join (
-                    select concept_cd, min(name_char) name
-                    from {2}.concept_dimension
+                    select concept_cd, min(name) name
+                    from (
+		      select c_basecode concept_cd,c_name name from {3}.{4}
+		      where ({5}) and {6}
+		      union all
+		      select concept_cd,name_char name from {2}.concept_dimension
+		      )
                     group by concept_cd
                 ) cd on cd.concept_cd = chicon.ccd
 
                 union all
                 select 'TOTAL' prefix, 'TOTAL' ccd, '' name
                 , (select count(distinct pn) from {1}) total
                 , 1 frc_total from dual
-                '''.format(pcounts, pconcepts, schema)
+                '''.format(pcounts, pconcepts, schema, self.metaschema, self.termtable, self.branchnodes, self.allbranchnodes)
                 cols, rows = do_log_sql(db, sql)
 
                 sql = '''
                 create index {0}_idx
                 on {0} (ccd)
                 '''.format(pcounts)
                 cols, rows = do_log_sql(db, sql)
+                # prefix is for filtering the output, so needs an index
+                sql = '''
+                create index {0}_pfx_idx
+                on {0} (prefix)
+                '''.format(pcounts)
+                cols, rows = do_log_sql(db, sql)
+                # total is used for filtering by threshold, so needs an index
+                sql = '''
+                create index {0}_tl_idx
+                on {0} (total)
+                '''.format(pcounts)
+                cols, rows = do_log_sql(db, sql)
+	    try:
+		log.debug('Checking if empirical schemes table exists...')
```

Comment:
Here is a good place to describe the unification of ABB's and AFB's approaches to output filtering. Both filter on the PREFIX column of the chi_pcounts table. ABB gets the allowed PREFIX values and corresponding human-readable names from the metaschema.SCHEMES table. AFB wanted to get them from the distinct values of PREFIX in the chi_pcounts table itself. The advantage of AFB's approach is that the user will never be offered selections that produce no results, and if a prefix exists in chi_pcounts it will not get omitted ( and results getting omitted is what precipitated the current round of changes in the first place). The advantage of ABB's approach is that the user is not presented with a menu of incomprehensible alphanumerics because we're also pulling names.

So below is the compromise: we create a chinotype-specific lookup table by pulling the PREFIXes from chi_pcounts (and rename them to C_NAME for compatibility with other functions that rely on it) and we left-join the C_KEY and C_DESCRIPTION values from the metaschema.SCHEMES table. Some prefixes will be missing from SCHEMES and so their C_KEY and C_DESCRIPTION values will remain NULL. We then go back and fill in the NULLs by copying over the C_NAME values into them. 

Yes, this is an extra table, but it's tiny and it fixes our problem. Heck, we can even use it as a diagnostic for what needs to be updated in the SCHEMES table.